File: HisabCore/Sources/HisabCore/SelfTransfers.swift
import Foundation

/// Detects bank→bank self transfers between the user's own accounts: a debit in one
/// bank source paired with an equal credit in a *different* bank source within the
/// date window. Both sides are internal movements — neither expense nor income.
public enum SelfTransfers {
    public static func detect(bank: [(txn: ReconTxn, source: Source)],
                              dateWindowDays: Int = 2) -> Set<UUID> {
        let window = TimeInterval(dateWindowDays) * 86_400
        let debits = bank.filter { $0.txn.direction == .debit }
        var credits = bank.filter { $0.txn.direction == .credit }
        var flagged = Set<UUID>()

        for debit in debits.sorted(by: { $0.txn.date < $1.txn.date }) {
            let candidates = credits.enumerated()
                .filter { _, credit in
                    credit.source != debit.source
                        && credit.txn.amountPaise == debit.txn.amountPaise
                        && abs(credit.txn.date.timeIntervalSince(debit.txn.date)) <= window
                }
                .sorted { lhs, rhs in
                    abs(lhs.element.txn.date.timeIntervalSince(debit.txn.date))
                        < abs(rhs.element.txn.date.timeIntervalSince(debit.txn.date))
                }
            if let (index, credit) = candidates.first {
                flagged.insert(debit.txn.id)
                flagged.insert(credit.txn.id)
                credits.remove(at: index)
            }
        }
        return flagged
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/self_transfers.dart
/// Cross-bank self-transfer detection. Port of SelfTransfers.swift.
library;

import 'domain.dart';
import 'reconciliation.dart';

class BankTxn {
  final ReconTxn txn;
  final Source source;
  const BankTxn(this.txn, this.source);
}

class SelfTransfers {
  /// A debit in one bank source paired with an equal credit in a *different*
  /// bank source within the date window. Both sides are internal movements.
  static Set<String> detect(List<BankTxn> bank, {int dateWindowDays = 2}) {
    final window = Duration(days: dateWindowDays);
    final debits = bank.where((b) => b.txn.direction == Direction.debit).toList()
      ..sort((a, b) => a.txn.date.compareTo(b.txn.date));
    final credits =
        bank.where((b) => b.txn.direction == Direction.credit).toList();
    final flagged = <String>{};

    for (final debit in debits) {
      BankTxn? best;
      var bestIndex = -1;
      Duration? bestDelta;
      for (var i = 0; i < credits.length; i++) {
        final credit = credits[i];
        if (credit.source == debit.source) continue;
        if (credit.txn.amountPaise != debit.txn.amountPaise) continue;
        final delta = credit.txn.date.difference(debit.txn.date).abs();
        if (delta > window) continue;
        if (bestDelta == null || delta < bestDelta) {
          best = credit;
          bestIndex = i;
          bestDelta = delta;
        }
      }
      if (best != null) {
        flagged.add(debit.txn.id);
        flagged.add(best.txn.id);
        credits.removeAt(bestIndex);
      }
    }
    return flagged;
  }
}
